File: app/models/project.py
from datetime import datetime
from enum import Enum

from sqlalchemy import DateTime, ForeignKey, String, UniqueConstraint, func
from sqlalchemy.orm import Mapped, mapped_column

from app.models.base import Base


class ProjectRole(str, Enum):
    owner = "owner"
    admin = "admin"
    member = "member"


class Project(Base):
    __tablename__ = "projects"

    id: Mapped[int] = mapped_column(primary_key=True, index=True)
    name: Mapped[str] = mapped_column(String(100), nullable=False)
    description: Mapped[str | None] = mapped_column(String(500), nullable=True)
    owner_id: Mapped[int] = mapped_column(ForeignKey("users.id"), nullable=False, index=True)
    invite_code: Mapped[str] = mapped_column(String(12), nullable=False, unique=True, index=True)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        server_default=func.now(),
        nullable=False,
    )


class ProjectMember(Base):
    __tablename__ = "project_members"
    __table_args__ = (UniqueConstraint("project_id", "user_id", name="uq_project_member"),)

    id: Mapped[int] = mapped_column(primary_key=True)
    project_id: Mapped[int] = mapped_column(ForeignKey("projects.id"), nullable=False, index=True)
    user_id: Mapped[int] = mapped_column(ForeignKey("users.id"), nullable=False, index=True)
    role: Mapped[str] = mapped_column(String(20), nullable=False, default=ProjectRole.member.value)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        server_default=func.now(),
        nullable=False,
    )

File: frontend/src/types/project.ts
export type ProjectRole = 'owner' | 'admin' | 'member'

export interface Project {
  id: number
  name: string
  description: string | null
  owner_id: number
  invite_code: string
  created_at: string
}

export interface ProjectListItem extends Project {
  role: ProjectRole
}

export interface ProjectMember {
  user_id: number
  username: string
  email: string
  role: ProjectRole
}

export interface ProjectMembership {
  project_id: number
  user_id: number
  role: ProjectRole
}

export interface CreateProjectPayload {
  name: string
  description: string | null
}

export interface JoinProjectPayload {
  projectId: number
  invite_code: string
}
